modules/fx/arrow-fx-kotlinx-coroutines/src/test/kotlin/arrow/integrations/kotlinx/CoroutinesIntegrationTest.kt

```diff
@@ -0,0 +1,166 @@
+import arrow.fx.typeclasses.seconds
+import arrow.test.UnitSpec
+import arrow.test.generators.throwable
+import io.kotlintest.fail
+import io.kotlintest.properties.Gen
+import io.kotlintest.properties.forAll
+import io.kotlintest.shouldBe
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.ObsoleteCoroutinesApi
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.newSingleThreadContext
+import kotlinx.coroutines.test.TestCoroutineDispatcher
+import kotlinx.coroutines.test.TestCoroutineExceptionHandler
+import kotlinx.coroutines.test.TestCoroutineScope
+
+@ObsoleteCoroutinesApi
+@Suppress("IMPLICIT_NOTHING_AS_TYPE_PARAMETER")
+@UseExperimental(ExperimentalCoroutinesApi::class)
+class CoroutinesIntegrationTest : UnitSpec() {
+
+  private val other = newSingleThreadContext("other")
+  private val all = newSingleThreadContext("all")
+
+  init {
+    // --------------- suspendCancellable ---------------
+
+    "suspendedCancellable should throw" {
+      forAll(Gen.throwable()) { e ->
+        val ceh = TestCoroutineExceptionHandler()
+        val scope = TestCoroutineScope(ceh + TestCoroutineDispatcher())
+
+        scope.launch {
+          IO { throw e }.suspendCancellable()
+        }
+        // suspendCancellableCoroutine re throws the exception so we need to compare the instance
+        ceh.uncaughtExceptions[0]::class == e::class
+      }
+    }
+
+    "suspendedCancellable should resume with right block" {
+      forAll(Gen.int()) { i ->
+        val ceh = TestCoroutineExceptionHandler()
+        val scope = TestCoroutineScope(ceh + TestCoroutineDispatcher())
+        scope.launch {
+          val first = IO { i + 1 }.suspendCancellable()
+          val second = IO { first + 1 }.suspendCancellable()
+          val third = IO { second + 1 }.suspendCancellable()
+
+          third shouldBe i + 3
+        }
+
+        ceh.uncaughtExceptions.isEmpty()
+      }
+    }
+
+    "scope cancellation should cancel suspendedCancellable IO" {
+      forAll(Gen.int()) { i ->
+        IO.fx {
+          val scope = TestCoroutineScope(Job() + TestCoroutineDispatcher())
+          val promise = !Promise<Int>()
+          !effect {
+            scope.launch {
+              IO.cancelable<Unit> { promise.complete(i) }.suspendCancellable()
+            }
+          }
+          !effect { scope.cancel() }
+          !promise.get()
+        }.unsafeRunSync() == i
+      }
+    }
+
+    "suspendCancellable can cancel even for infinite asyncs" {
+      IO.async { cb: (Either<Throwable, Int>) -> Unit ->
+        val scope = TestCoroutineScope(Job() + TestCoroutineDispatcher())
+        scope.launch {
+          IO.async<Int> { Thread.sleep(5000) }
+            .onCancel(IO { cb(1.right()) })
+            .suspendCancellable().let {
+              cb(it.right())
+            }
+        }
+        IO(other) { Thread.sleep(500) }
```

Comment:
```suggestion
        IO.sleep(500.milliseconds)
```

Never use `Thread.sleep` always use `IO.sleep` or other semantically blocking calls.

Since coroutines simply wrap Java's thread pools using `Thread.sleep` can block multiple tasks from happening since it blocks the Thread. Since commonly threads are shared in work-stealing pools, this means you're breaking or violating the semantics of a work-stealing pool. That can drastically alter the behavior of your program and not realistically mimic the workings of an actual program working on a work-stealing thread.

=> `Thread.sleep` is the equivalent of running blocking Java tasks such as JBDC calls on a work-stealing pool which is a huge anti-pattern.

Reply:
Yes, agreed. I'm not sure how it ended like this, it might have been a copy+paste problem from another test or something we changed during the process 😅